src/world_generator.h

```diff
@@ -277,25 +304,323 @@ namespace tobor {
+							if (world.blocked(cell_id) && (selected_index >= cell_id)) {
+								++selected_index;
+							}
+						}
+					}
+
+					for (typename world_type::int_type cell_id = 0; cell_id < BOARD_SIZE; ++cell_id) {
+						for (auto& selected_index : selected_indices_non_target) {
+							bool is_blocked =
+								world.blocked(cell_id)
+								||
+								std::find(selected_indices_target.cbegin(), selected_indices_target.cend(), cell_id) != selected_indices_target.cend();
+
+							if (is_blocked && (selected_index >= cell_id)) {
+								++selected_index;
+							}
+						}
+					}
+
+					using target_pieces_array_type = typename positions_of_pieces_type::target_pieces_array_type;
+					using non_target_pieces_array_type = typename positions_of_pieces_type::non_target_pieces_array_type;
+
+					target_pieces_array_type target_positioning;
+					for (std::size_t i = 0; i < selected_indices_target.size(); ++i) {
+						target_positioning[i] = cell_id_type::create_by_id(selected_indices_target[i], world);
+					}
+
+					non_target_pieces_array_type non_target_positioning;
+					for (std::size_t i = 0; i < selected_indices_non_target.size(); ++i) {
+						non_target_positioning[i] = cell_id_type::create_by_id(selected_indices_non_target[i], world);
+					}
+
+					auto result = positions_of_pieces_type(target_positioning, non_target_positioning);
+
+					return result;
+				}
+
+
+				inline initial_state_generator& operator++() {
+					++counter;
+					counter %= CYCLIC_GROUP_SIZE;
+					return *this;
+				}
+
+				inline initial_state_generator& operator--() {
+					counter += CYCLIC_GROUP_SIZE - 1; // check overflow (?)
+					counter %= CYCLIC_GROUP_SIZE;
+					return *this;
+				}
+
+				void set_counter(const uint64_t& counter_p) {
+					counter = counter_p % CYCLIC_GROUP_SIZE;
+				}
 			};
 
+			template<class Main_Group_Generator_Type, class Side_Group_Generator_Type>
+
 			class product_group_generator {
 
+			private:
+
+				static constexpr decltype(Side_Group_Generator_Type::CYCLIC_GROUP_SIZE)
+					MAIN_OVERFLOW_SIDE_DIFFERENCE{ Main_Group_Generator_Type::CYCLIC_GROUP_SIZE % Side_Group_Generator_Type::CYCLIC_GROUP_SIZE };
+
+				Main_Group_Generator_Type main_generator;
+				Side_Group_Generator_Type side_generator;
+
+			public:
+
+				using main_group_generator_type = Main_Group_Generator_Type;
+
+				using side_group_generator_type = Side_Group_Generator_Type;
+
+				product_group_generator() {
+				}
+
+				Main_Group_Generator_Type& main() {
+					return main_generator;
+				}
+
+				Side_Group_Generator_Type& side() {
+					return side_generator;
+				}
+
+				// add counter / generator here if wanted. for random access
+
+
 				// main Group generator size M
 
 				// side Group generator size N
 
 				// generator standard 1
-				
+
 				// counter k in 0..M*N
 
 				// in main group select k mod M
 
-				// in side group select (k mode M + k div M ) mod N
+				// in side group select (k mod M + k div M ) mod N
```

Comment:
add at least two concepts of generator:
* random access generator
* bidirectional generator
* forward generator

Use big ints to implement a random access generator instead of this variant.
Check fo rall generators that there is no overflow.

Reply:
to be done within #111